cirq/circuits/circuit.py

```diff
@@ -1387,6 +1390,126 @@ def tetris_concat(
+        for op in self.all_operations():
+            if len(op.qubits) > 1:
+                uf.union(*op.qubits)
+        return sorted([qs for qs in uf.to_sets()], key=min)
+
+    def factorize(self: CIRCUIT_TYPE) -> Iterable[CIRCUIT_TYPE]:
+        """Factorize circuit into a sequence of independent circuits (factors).
+
+        Factorization is possible when the circuit's qubits can be divided
+        into two or more independent qubit sets. Preserves the moments from
+        the original circuit.
+        If this is not possible, returns the set consisting of the single
+        circuit (this one).
+
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> circuit = cirq.Circuit()
+        >>> circuit.append(cirq.Moment(cirq.H(q2)))
+        >>> circuit.append(cirq.Moment(cirq.CZ(q0,q1)))
+        >>> circuit.append(cirq.H(q0))
+        >>> print(circuit)
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        <BLANKLINE>
+        2: ───H───────────
+        >>> for i, f in enumerate(circuit.factorize()):
+        ...     print("Factor {}".format(i))
+        ...     print(f)
+        ...
+        Factor 0
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        Factor 1
+        2: ───H───────────
+
+        Returns:
+            The sequence of circuits, each including only the qubits from one
+            independent qubit set.
+
+        """
+
+        qubit_factors = self.get_independent_qubit_sets()
+        if len(qubit_factors) == 1:
+            return (self,)
+        return (
+            self._with_sliced_moments([m[qubits] for m in self.moments]) for qubits in qubit_factors
+        )
+
+    def slice_by_qubits(self: CIRCUIT_TYPE, with_qubits: Collection['cirq.Qid']) -> CIRCUIT_TYPE:
+        """Retuns part of the circuit entangled with the given qubits.
+
+        If a circuit can be split into independent factors, that means only
+        returning the factors which contain at least one of the given qubits.
+
+        If a circuit cannot be factorized into independent qubit sets, returns
+        the whole circuit.
+
+        >>> q0, q1, q2 = cirq.LineQubit.range(3)
+        >>> circuit = cirq.Circuit()
+        >>> circuit.append(cirq.Moment(cirq.H(q2)))
+        >>> circuit.append(cirq.Moment(cirq.CZ(q0,q1)))
+        >>> circuit.append(cirq.H(q0))
+        >>> print(circuit)
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+        <BLANKLINE>
+        2: ───H───────────
+        >>> print(circuit.slice_by_qubits([q1]))
+        0: ───────@───H───
+                  │
+        1: ───────@───────
+
+        Args:
+            with_qubits: Qubits for which the 'circuit slice' should be created.
+
+        Returns:
+            The circuit, including only the qubits which are entangled with the
+            provided qubits.
+
+        """
+
+        if not frozenset(with_qubits) < self.all_qubits():
```

Comment:
should this be `<=`? Instead of `not <` would it be easier to read as `>`?

Reply:
Yeah. Anyways, gone now :)